subscription_plan.go

```diff
@@ -0,0 +1,226 @@
+		},
+		{
+			Operation: "replace",
+			Path:      "/payment_preferences/auto_bill_outstanding",
+			Value:     self.PaymentPreferences.AutoBillOutstanding,
+		},
+		{
+			Operation: "replace",
+			Path:      "/payment_preferences/payment_failure_threshold",
+			Value:     self.PaymentPreferences.PaymentFailureThreshold,
+		},
+		{
+			Operation: "replace",
+			Path:      "/payment_preferences/setup_fee_failure_action",
+			Value:     self.PaymentPreferences.SetupFeeFailureAction,
+		},
+		{
+			Operation: "replace",
+			Path:      "/taxes/percentage",
+			Value:     self.Taxes.Percentage,
+		},
+	}
+
+	if self.PaymentPreferences.SetupFee != nil {
+		result = append(result, Patch{
+			Operation: "replace",
+			Path:      "/payment_preferences/setup_fee",
+			Value:     self.PaymentPreferences.SetupFee,
+		},
+		)
+	}
+
+	return result
+}
+
+// CreateSubscriptionPlan creates a subscriptionPlan
+// Doc: https://developer.paypal.com/docs/api/subscriptions/v1/#plans_create
+// Endpoint: POST /v1/billing/plans
+func (c *Client) CreateSubscriptionPlan(newPlan SubscriptionPlan) (*CreateSubscriptionPlanResponse, error) {
+	req, err := c.NewRequest(http.MethodPost, fmt.Sprintf("%s%s", c.APIBase, "/v1/billing/plans"), newPlan)
+	response := &CreateSubscriptionPlanResponse{}
+	if err != nil {
+		return response, err
+	}
+	err = c.SendWithAuth(req, response)
+	return response, err
+}
+
+// UpdateSubscriptionPlan. updates a plan
+// Doc: https://developer.paypal.com/docs/api/subscriptions/v1/#plans_patch
+// Endpoint: PATCH /v1/billing/plans/:plan_id
+func (c *Client) UpdateSubscriptionPlan(updatedPlan SubscriptionPlan) error {
+	req, err := c.NewRequest(http.MethodPatch, fmt.Sprintf("%s%s%s", c.APIBase, "/v1/billing/plans/", updatedPlan.ID), updatedPlan.GetUpdatePatch())
+	if err != nil {
+		return err
+	}
+	err = c.SendWithAuth(req, nil)
+	return err
+}
+
+// UpdateSubscriptionPlan. updates a plan
+// Doc: https://developer.paypal.com/docs/api/subscriptions/v1/#plans_get
+// Endpoint: GET /v1/billing/plans/:plan_id
+func (c *Client) GetSubscriptionPlan(planId string) (*SubscriptionPlan, error) {
+	req, err := c.NewRequest(http.MethodGet, fmt.Sprintf("%s%s%s", c.APIBase, "/v1/billing/plans/", planId), nil)
+	response := &SubscriptionPlan{}
+	if err != nil {
+		return response, err
+	}
+	err = c.SendWithAuth(req, response)
+	return response, err
+}
+
+// List all plans
+// Doc: https://developer.paypal.com/docs/api/subscriptions/v1/#plans_list
+// Endpoint: GET /v1/billing/plans
+func (c *Client) ListSubscriptionPlans(params *SubscriptionPlanListParameters) (*ListProductsResponse, error) {
+	req, err := c.NewRequest(http.MethodGet, fmt.Sprintf("%s%s", c.APIBase, "/v1/billing/plans"), nil)
+	response := &ListProductsResponse{}
+	if err != nil {
+		return response, err
+	}
+
+	if params != nil {
+		q := req.URL.Query()
+		q.Add("page", params.Page)
+		q.Add("page_size", params.PageSize)
+		q.Add("total_required", params.TotalRequired)
```

Comment:
is it worth extracting this bit to a method?

Reply:
This must go into a separate refactoring branch where this is changed everywhere